<a href="https://colab.research.google.com/github/terrafirmatrekker/WHR2022EDA/blob/main/WorldHappinessReport22EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Analyzing The World Happiness Report of 2022**

2022 is the tenth anniversary of the World Happiness
Report (WHR). The WHRs are based
on a wide variety of data, the most important
source has always been the Gallup World Poll, due to its
unique range and comparability of global
annual surveys.

Evaluations of quality of life from the Gallup World Poll
provides the basis for the annual happiness
rankings that have always sparked widespread
interest about the secrets of life in the happiest
countries. 

## Methodology

The Gallup World poll used asks respondents to evaluate their
current life as a whole using the mental image of a ladder, with **the best possible life for them as a 10 and worst possible as a 0.** 

Each respondent provides a numerical responseon this scale, referred to as the Cantril ladder. Typically, around 1,000 responses are gathered annually for each country. Weights are used to construct population-representative national averages for each year in each country. We base our national happiness rankings on a three-year average, thereby increasing the
sample size to provide more precise estimates. 

## Data

This report will examine the most recent [dataset](https://happiness-report.s3.amazonaws.com/2022/Appendix_2_Data_for_Figure_2.1.xls) which has been converted into a CSV and stored as a Gist [here](https://gist.github.com/terrafirmatrekker/474f3acc5b44322bc54fcc49870dcfd1).:

## Loading and Cleaning Data

In [4]:
# Importing Some of the Initial Libraries Needed
import pandas as pd
from urllib.request import urlretrieve

In [14]:
url = 'https://gist.githubusercontent.com/terrafirmatrekker/474f3acc5b44322bc54fcc49870dcfd1/raw/14dafbd7faf09e0b1d4a53646b830cf79b3c510b/WHR2022.csv'
urlretrieve(url, 'WHR2022.csv')
data = pd.read_csv('WHR2022.csv')
# Get a count of unique values and columns
print(data.shape)
# Get top 6 countries
data.head(6)

(147, 24)


RANK      Country  Happiness score  Whisker-high  Whisker-low  \
0     1      Finland            7.821         7.886        7.756   
1     2      Denmark            7.636         7.710        7.563   
2     3      Iceland            7.557         7.651        7.464   
3     4  Switzerland            7.512         7.586        7.437   
4     5  Netherlands            7.415         7.471        7.359   
5     6  Luxembourg*            7.404         7.501        7.307   

   Dystopia (1.83) + residual  Explained by: GDP per capita  \
0                       2.518                         1.892   
1                       2.226                         1.953   
2                       2.320                         1.936   
3                       2.153                         2.026   
4                       2.137                         1.945   
5                       2.042                         2.209   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                         1.258                                  0.775   
1                         1.243                                  0.777   
2                         1.320                                  0.803   
3                         1.226                                  0.822   
4                         1.206                                  0.787   
5                         1.155                                  0.790   

   Explained by: Freedom to make life choices  ...  Unnamed: 14  Unnamed: 15  \
0                                       0.736  ...          NaN          NaN   
1                                       0.719  ...          NaN          NaN   
2                                       0.718  ...          NaN          NaN   
3                                       0.677  ...          NaN          NaN   
4                                       0.651  ...          NaN          NaN   
5                                       0.700  ...          NaN          NaN   

   Unnamed: 16  Unnamed: 17  Unnamed: 18  Unnamed: 19  Unnamed: 20  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   
5          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 21  Unnamed: 22  Unnamed: 23  
0          NaN          NaN          NaN  
1          NaN          NaN          NaN  
2          NaN          NaN          NaN  
3          NaN          NaN          NaN  
4          NaN          NaN          NaN  
5          NaN          NaN          NaN  

[6 rows x 24 columns]

As we can see there are 24 columns, but it seems like some of the columns are empty. Let's use *pandas.DataFrame.info* method to get more info about the data types.


In [16]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 147 entries, 0 to 146
Data columns (total 24 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   RANK                                        147 non-null    int64  
 1   Country                                     147 non-null    object 
 2   Happiness score                             146 non-null    float64
 3   Whisker-high                                146 non-null    float64
 4   Whisker-low                                 146 non-null    float64
 5   Dystopia (1.83) + residual                  146 non-null    float64
 6   Explained by: GDP per capita                146 non-null    float64
 7   Explained by: Social support                146 non-null    float64
 8   Explained by: Healthy life expectancy       146 non-null    float64
 9   Explained by: Freedom to make life choices  146 non-null    float64
 10  Explained by: 

So it seems as if we have 12 empty columns without values that we can remove prior to our analysis. Let's us *df.dropna(axis=1, how='all')* to remove these extraneous columns with NaNs and store it in a new dataframe.

In [27]:
clean_data = data.dropna(axis=1, how='all')
clean_data.head()

RANK      Country  Happiness score  Whisker-high  Whisker-low  \
0     1      Finland            7.821         7.886        7.756   
1     2      Denmark            7.636         7.710        7.563   
2     3      Iceland            7.557         7.651        7.464   
3     4  Switzerland            7.512         7.586        7.437   
4     5  Netherlands            7.415         7.471        7.359   

   Dystopia (1.83) + residual  Explained by: GDP per capita  \
0                       2.518                         1.892   
1                       2.226                         1.953   
2                       2.320                         1.936   
3                       2.153                         2.026   
4                       2.137                         1.945   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                         1.258                                  0.775   
1                         1.243                                  0.777   
2                         1.320                                  0.803   
3                         1.226                                  0.822   
4                         1.206                                  0.787   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                       0.736                     0.109   
1                                       0.719                     0.188   
2                                       0.718                     0.270   
3                                       0.677                     0.147   
4                                       0.651                     0.271   

   Explained by: Perceptions of corruption  
0                                    0.534  
1                                    0.532  
2                                    0.191  
3                                    0.461  
4                                    0.419

It seems that our data has two other columns that we can drop since "Whisker-high" & "Whisker-low" refers to a box plot that was generated from the orginal dataset and is not necessary for analysis so let's drop these columns with *DataFrame.drop*.

In [30]:
clean_data = clean_data.drop(labels=['Whisker-high', 'Whisker-low'], axis = 1)
clean_data.head()

RANK      Country  Happiness score  Dystopia (1.83) + residual  \
0     1      Finland            7.821                       2.518   
1     2      Denmark            7.636                       2.226   
2     3      Iceland            7.557                       2.320   
3     4  Switzerland            7.512                       2.153   
4     5  Netherlands            7.415                       2.137   

   Explained by: GDP per capita  Explained by: Social support  \
0                         1.892                         1.258   
1                         1.953                         1.243   
2                         1.936                         1.320   
3                         2.026                         1.226   
4                         1.945                         1.206   

   Explained by: Healthy life expectancy  \
0                                  0.775   
1                                  0.777   
2                                  0.803   
3                                  0.822   
4                                  0.787   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                       0.736                     0.109   
1                                       0.719                     0.188   
2                                       0.718                     0.270   
3                                       0.677                     0.147   
4                                       0.651                     0.271   

   Explained by: Perceptions of corruption  
0                                    0.534  
1                                    0.532  
2                                    0.191  
3                                    0.461  
4                                    0.419

## Exploring the Data

Now that we have cleaned the data for our analysis. Let's first use *pandas.DataFrame.describe* to get a quick snapshot of some descriptive statistics. 

In [35]:
clean_data.describe()

RANK  Happiness score  Dystopia (1.83) + residual  \
count  147.000000       146.000000                  146.000000   
mean    74.000000         5.553575                    1.831808   
std     42.579338         1.086843                    0.534994   
min      1.000000         2.404000                    0.187000   
25%     37.500000         4.888750                    1.555250   
50%     74.000000         5.568500                    1.894500   
75%    110.500000         6.305000                    2.153000   
max    147.000000         7.821000                    2.844000   

       Explained by: GDP per capita  Explained by: Social support  \
count                    146.000000                    146.000000   
mean                       1.410445                      0.905863   
std                        0.421663                      0.280122   
min                        0.000000                      0.000000   
25%                        1.095500                      0.732000   
50%                        1.445500                      0.957500   
75%                        1.784750                      1.114250   
max                        2.209000                      1.320000   

       Explained by: Healthy life expectancy  \
count                             146.000000   
mean                                0.586171   
std                                 0.176336   
min                                 0.000000   
25%                                 0.463250   
50%                                 0.621500   
75%                                 0.719750   
max                                 0.942000   

       Explained by: Freedom to make life choices  Explained by: Generosity  \
count                                  146.000000                146.000000   
mean                                     0.517226                  0.147377   
std                                      0.145859                  0.082799   
min                                      0.000000                  0.000000   
25%                                      0.440500                  0.089000   
50%                                      0.543500                  0.132500   
75%                                      0.626000                  0.197750   
max                                      0.740000                  0.468000   

       Explained by: Perceptions of corruption  
count                               146.000000  
mean                                  0.154781  
std                                   0.127514  
min                                   0.000000  
25%                                   0.068250  
50%                                   0.119500  
75%                                   0.198500  
max                                   0.587000

This method helped us get an idea of the central tendency, dispersion and shape of the dataset’s distribution. Here we can see that the **Mean Happiness Score** is 5.553575 & the Min and Max are 2.4 and 7.8 respectively. 

Let's create a few functions to highlight  the maximum values and minimum values of each attribute in the data set.

In [46]:
# Highlight the Maximum values
def highlight_max(t):    
    is_max = t == t.max()
    return ['background-color: green' if v else '' for v in is_max]

clean_data.style.apply(highlight_max)

In [45]:
# Highlight the Minimum Values
def highlight_min(b):
  is_min = b == b.min()
  return ['background-color: coral' if v else '' for v in is_min]

clean_data.style.apply(highlight_min)